Add tests via pytest, asynctest

tslaton opened the pull request:
Currently these tests do not necessarily represent server behavior faithfully, but they do achieve 100% coverage of `sshapiauthenticator/auth.py`. 

They also demonstrate one straightforward means of mocking out asynchronous network code.

There may be a cleaner method of generating a valid ssh key without calling `ssh-keygen` via `subprocess`. Although the method currently used will produce temporary files on the user's system, it also removes them.

tslaton committed:
Add tests via pytest, asynctest

scanon commented on tests/test_authenticator.py:
```diff
@@ -0,0 +1,97 @@
+    authenticator.skey = 'non-empty'
+    data = {'username': 'tester', 'password': 'valid'}
+    resp = await authenticator.authenticate(None, data)
+    assert resp == 'tester'
```
Maybe add some tests to confirm that the ssh files were created.

scanon replied:
Do this for all the successful tests.

scanon commented on tests/test_authenticator.py:
```diff
@@ -0,0 +1,97 @@
+    handler = Mock(**{'request': Mock(**{'remote_ip': 'http://localhost:9998'})})
+    data = {'username': 'tester', 'password': 'invalid'}
+    resp = await authenticator.authenticate(handler, data)
+    assert resp is None
```
If we add logging, then you could mock the log call and confirm that some log message is being generated.

scanon reviewed (changes requested):
This is looking good.  I think add a few more tests and some logging and it is good for a first cut.

tslaton committed:
tests: Add checks for key creation and logs

tslaton commented:
I added checks for the file creation and some log messages. I'm still not 100% sure I'm logging the best messages.

I added a bit about the response error code and human readable reason to the logs in the case where all of that is defined. 

Also, I made `SSHAPIAuthenticator` use its encoding trait, but I haven't tested any non `utf8` encodings.

scanon reviewed (approved):
This all looks good.
